Add rumantsch

manuelmeister opened the pull request:
I've created a draft based on #4013 and #3958

Closes #4013
Closes #3958

eruedin committed:
Create rm-CH-scout.json

new, rm

eruedin committed:
Create rm.json

Translated by Martin Cantieni from dicziunari.ch (former journalist at RTR and board member at Pro Svizra Rumantscha)

manuelmeister committed:
Move rm to frontend locales

manuelmeister committed:
Merge remote-tracking branch 'upstream/devel' into feature/rumantsch

manuelmeister committed:
Adjust language fallback mechanism

manuelmeister committed:
Add missing return

manuelmeister committed:
Move return to end

DeNic0la commented:
Hey, hat bei mir nicht funktioniert (ist nur technisches Feedback, nicht zu den Übersetzungen).
api/src/Entity/Profile.php zeile 160: Die Option ist nicht zugelassen.

Dies wirft im Frontend den folgenden Fehler auf:
![image](https://github.com/ecamp/ecamp3/assets/57986114/ebd118a9-c885-4e44-b59e-d91965643895)

manuelmeister commented:
@DeNic0la okay. In diesem PR geht es mir mal nur um das Frontend auszuprobieren. Weniger um das gerade zu releasen.

manuelmeister committed:
Add rumantsch choice to profile endpoint

manuelmeister committed:
Update api snapshot to include rumantsch

manuelmeister commented:
I fixed the errors. 
@ecamp/core how should we continue here? Should we merge this and so that rumantsch can be improved on lokalise?

manuelmeister commented:
## Core Meeting Decision

Open for review by team

carlobeltrame commented on frontend/src/locales/rm.json:
```diff
@@ -0,0 +1,625 @@
+    },
+    "serverError": {
+      "409": "Uuuups... Questa acziun ha chaschunà ina errur sin il server.",
+      "short": "Serverfehler"
```
Remove this clearly German string?

BacLuc replied:
done

carlobeltrame commented on frontend/src/plugins/i18n/apiFallbackLocalesFor.js:
```diff
@@ -1,13 +1,34 @@
-  const parts = locale.split('_')
-  for (let i = parts.length - 1; i > 0; i--) {
-    yield parts.slice(0, i).join('_')
+  if (typeof fallbackLocale === 'string') {
+    yield fallbackLocale
+  } else if (Array.isArray(fallbackLocale)) {
+    for (const fallback of fallbackLocale) {
+      yield fallback
+    }
+  } else {
```
Why do we want to start supporting string and Array fallback locales? You are importing `fallbackLocale` from a single fixed file in our repo. Is this a case of YAGNI?

BacLuc replied:
done

manuelmeister replied:
But isn't this the usecase for rumantsch where you want german as a fallback of `rm-CH-scout`? Otherwise it will just be `rm-CH` and `rm`

carlobeltrame commented on frontend/src/plugins/i18n/index.js:
```diff
@@ -109,7 +116,20 @@ Object.defineProperty(i18n, 'browserPreferredLocale', {
       }
     }
-    return fallbackLocale
+    if (typeof fallbackLocale === 'string') {
+      return fallbackLocale
+    } else if (Array.isArray(fallbackLocale)) {
+      for (const fallback of fallbackLocale) {
+        return fallback
+      }
```
Same here

BacLuc replied:
done

manuelmeister replied:
This is not the same use case. Here you just select the desired language based on browser locale.

carlobeltrame commented on frontend/src/plugins/i18n/index.js:
```diff
@@ -109,7 +116,20 @@ Object.defineProperty(i18n, 'browserPreferredLocale', {
+        return fallback
+      }
+    } else {
+      if ('default' in fallbackLocale) {
```
Is this condition ever met? `'default'` is not in our current `fallbackLocale` object as far as I can see, only `['default']` is..?

BacLuc replied:
done

BacLuc commented:
This leads to an error when resetting the password:
`{"message":"Uncaught PHP Exception Exception: \"Can not find Mail-Template translated 'emails/passwordResetLink.{language}.html.twig' for rm-CH-scout\" at /srv/api/src/Service/MailService.php line 134","context":{"exception":{"class":"Exception","message":"Can not find Mail-Template translated 'emails/passwordResetLink.{language}.html.twig' for rm-CH-scout","code":0,"file":"/srv/api/src/Service/MailService.php:134"}},"level":500,"level_name":"CRITICAL","channel":"request","datetime":"2023-11-10T21:17:51.102528+00:00","extra":{}}`

BacLuc committed:
Merge remote-tracking branch 'upstream/devel' into feature/rumantsch

BacLuc committed:
api: wrap the langauges in Profile

The line was quite long now.

BacLuc committed:
api: update ResponseSnapshotTest__testOpenApiSpecMatchesSnapshot__1.yml

BacLuc committed:
api: add readonly modifiers to constructor parameters in MailService

They were introduced in php 8.1.

BacLuc committed:
api: configure fallback for rm in MailService

BacLuc committed:
api: add possibility to translate validation messages to rumantsch

BacLuc committed:
api: use english as global fallback in MailService

BacLuc committed:
frontend: revert complicated fallback logic

BacLuc committed:
frontend: use fallback logic from vue-i18n to fallback to de from rm

As documented here: https://kazupon.github.io/vue-i18n/guide/fallback.html#explicit-fallback-with-decision-maps
Reimplement the vue-i18n fallback logic for the api error message
translation in apiFallbackLocalesFor.js
Add empty rumantsch translations to print and common,
and test that the translation works in print with one
rumantsch translation in common.

BacLuc committed:
frontend: remove german text from rumantsch translation

BacLuc commented:
@manuelmeister can you give a :+1: if you are happy with the changes.
Then we can merge this.
